Add County field in Registration and Address Edit forms

ankitdwivedi23 opened the pull request:
Closes issue #5

Commit:
AddCounty migration

Commit:
add county field in Registration form

Commit:
add county field to PickupLocation

Commit:
Fix code indentation

Commit:
Enable editing county

Commit:
Update ViewBag with list of states and counties whenever Grower/FoodBank View is returned

Commit:
Use AddressEditViewModel in FoodBankEditViewModel

Commit:
Rename RoleEditViewModel and use AddressEditViewModel

Commit:
Use shared partial view for address

alimon808 commented on NWHarvest.Web/Controllers/AccountController.cs:
```diff
@@ -563,7 +563,68 @@ private void RegisterViewBag()
+            List<string> counties =
+                new List<string>
+                {
+                    "Unknown",
```
Is 'Unknown' necessary?

ankitdwivedi23 replied:
Rachel thought it will be good to have this option in case Growers don't know what county they are in.

alimon808 replied:
I see.  If it's necessary for user to have this option rather than have them look it up, I suggest placing it as the last item on the list.  This will force those users who are unsure to go through the list before selecting Unknown.  Since the county field will eventually be used as one way to filter listings, user with unknown county will be an issue.

ankitdwivedi23 replied:
Makes sense. Update the list to have "Unknown" as the last option.

alimon808 commented on NWHarvest.Web/Migrations/201711100017429_StateAndCountyModels.cs:
```diff
@@ -0,0 +1,42 @@
+namespace NWHarvest.Web.Migrations
```
Storing a single state and related counties might be overkill in db.  Propose creating WashingtonState class instead.

ankitdwivedi23 replied:
I thought of that as well. But the db approach feels better for making the website open to users outside Washington. Although that will still require a change in how the County drop down gets updated. Let me know if you still feel we should go with the Class approach, and I will make that change.

alimon808 replied:
Yea, I prefer going with class approach using a StateViewModel that defaults to Washington state values.  This model can easily extended using another state by overriding the default values via in-memory or db.

I'm contemplating whether we need a State form field at all.  Since one of the requirements is that all users must be within Washington, we can implicitly add it when updating the db.  One less thing the user have to see when registering.

ankitdwivedi23 replied:
Sure. I updated the migration. About the state form field, I agree it isn't needed as per current requirements and we should remove it. But let's do that in a separate pull request.

alimon808 commented:
Cleaning up history by squashing both 'Fix code indentation'

alimon808 commented on NWHarvest.Web/Controllers/AccountController.cs:
```diff
@@ -554,16 +556,25 @@ private void RegisterViewBag()
-            };
+                    Text = s.Name,
+                    Value = s.ShortName,
+                    Selected = false
```
Set selected to true

alimon808 reviewed (changes requested):
Looks good.  See comments for requested changes.

Commit:
Set Washington as the selected state in registration form

ankitdwivedi23 commented:
Addressed comments. About squashing the "Fix code indentation" commits, I saw that after pulling new changes from master. I'm unsure if squashing them now will rewrite the master branch commits as well. Asked [this question](https://stackoverflow.com/questions/47362667/git-squash-commits-after-pulling-other-commits-from-master-branch) on SO. Will appreciate your help!

alimon808 commented:
It should be ok to rebase locally and then update the pull request.

Commit:
Fetch states and counties from DB

Commit:
Use separate partial view for editing Pickup Location address

Commit:
Populate State and Counties from WashingtonState class instead of DB

ankitdwivedi23 commented:
Rebased onto master and addressed remaining comments.

alimon808 commented on NWHarvest.Web/Controllers/PickupLocationsController.cs:
```diff
@@ -93,7 +93,8 @@ public ActionResult Create()
                     },
                     Address = new AddressEditViewModel
                     {
```
Unable to change county field for a pickup location.

ankitdwivedi23 replied:
I made it read-only intentionally, as we didn't want a pickup location county to be different from the grower's county. But I forgot to handle the scenario when a grower edits their county (what happens to the counties of existing pickup locations?)

To keep things simple, I have enabled editing county for pickup locations. We can discuss about this scenario and implement any necessary validation in a separate pull request.

alimon808 commented:
1) Unable to change the county field as grower.
2) Editing a Foodbank throws an exception:
The ViewData item that has the key 'County' is of type 'System.String' but must be of type 'IEnumerable<SelectListItem>'.

Commit:
Resolve conflict

Commit:
Enable editing pickup locations

Commit:
Recreate AddCounty migration and update Seed method

Commit:
Update PickupLocation state and county in DB on POST

Commit:
Call RegisterViewData() to get list of states and counties for FoodBank Edit form

ankitdwivedi23 commented:
Addressed comments. Thanks for finding the Foodbank edit exception!
